rubicon/java/android_events.py

```diff
@@ -168,45 +193,125 @@ def _run_scheduled_callbacks_then_restart_loop(self):
+    # in the app freezing, triggering an App Not Responding pop-up from the
+    # platform, and the user killing the app.
+    #
+    # Instead, the AndroidEventLoop cooperates with the native Android event
+    # loop to be woken up to get work done as needed.
+    def select(self, *args, **kwargs):
+        raise NotImplementedError(
+            "_SelectorMinusSelect refuses to select(); see comments."
         )
-        asyncio.events._set_running_loop(self)
 
-        # Create Android handler to run our own code on. We assume that we are running
-        # in the main Android activity thread, and we therefore do not worry about thread safety.
-        Handler = JavaClass("android/os/Handler")
-        self._handler = Handler()  # No-arg constructor uses main Android event loop.
-        # Create a Java Runnables to tick the event loop forward. We statically create
-        # Runnables to avoid creating lots of temporary objects.
-        python_self = self
 
-        class RunScheduledCallbacksThenRestartLoopRunnable(Runnable):
-            def run(self):
-                python_self._run_scheduled_callbacks_then_restart_loop()
+class AndroidInterop:
+    """Encapsulate details of Android event loop cooperation."""
+
+    def __init__(self):
+        self._runnable_by_key = {}
+        self._calling_soon = set()
+        self._handler = None
+
+    @property
+    def handler(self):
+        if self._handler is None:
+            # We use `android.os.Handler.postDelayed()` to ask for wakeup. This requires an instance
+            # of `Handler`, which we cache. We use the default constructor which assumes we are on
+            # the Android UI thread.
+            self._handler = Handler()
+        return self._handler
+
+    def get_or_create_runnable(self, key, fn):
+        if key in self._runnable_by_key:
+            return self._runnable_by_key[key]
+        android_interop_self = self
 
-        class RunEventLoopOnceCooperativelyRunnable(Runnable):
+        class PythonRunnable(Runnable):
             def run(self):
-                python_self._run_event_loop_once_cooperatively()
-
-        self._runnables = {
-            "_run_scheduled_callbacks_then_restart_loop": RunScheduledCallbacksThenRestartLoopRunnable(),
-            "_run_event_loop_once_cooperatively": RunEventLoopOnceCooperativelyRunnable(),
-        }
-        self._run_event_loop_once_cooperatively()
-        print("Started event loop")
+                if key in android_interop_self._calling_soon:
+                    android_interop_self._calling_soon.remove(key)
+                fn()
+
+        self._runnable_by_key[key] = PythonRunnable()
+        return self._runnable_by_key[key]
+
+    def call_soon_dedup(self, key, fn):
+        """Enqueue a Python callable `fn` to be run soon. Use `key` to identify avoid duplication:
+        if `call_soon_dedup()` is called twice with the same `key`, we only ask Android for one
```

Comment:
What *is* the key? A timestamp? A text label? Guidance on what to use would be helpful. 

Is there any way we can use an inherent property of the callable, rather than a manual key (e.g. the hash of the callable)?